pkg/api/openshiftcluster_validatestatic.go

```diff
@@ -0,0 +1,319 @@
+package api
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"github.com/Azure/ARO-RP/pkg/util/immutable"
+	"net"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/subnet"
+)
+
+const (
+	resourceProviderNamespace = "Microsoft.RedHatOpenShift"
+	resourceType              = "openShiftClusters"
+)
+
+var (
+	rxResourceGroupID = regexp.MustCompile(`(?i)^/subscriptions/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/resourceGroups/[-a-z0-9_().]{0,89}[-a-z0-9_()]$`)
+	rxSubnetID        = regexp.MustCompile(`(?i)^/subscriptions/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/resourceGroups/[-a-z0-9_().]{0,89}[-a-z0-9_()]/providers/Microsoft\.Network/virtualNetworks/[-a-z0-9_.]{2,64}/subnets/[-a-z0-9_.]{2,80}$`)
+	rxDomainName      = regexp.MustCompile(`^` +
+		`([a-z0-9]|[a-z0-9][-a-z0-9]{0,61}[a-z0-9])` +
+		`(\.([a-z0-9]|[a-z0-9][-a-z0-9]{0,61}[a-z0-9]))*` +
+		`$`)
+)
+
+func NewOpenShiftClusterStaticValidator(env env.Interface, resourceID string, converter OpenShiftClusterConverter) OpenShiftClusterStaticValidator {
+	return &openShiftClusterStaticValidator{
+		location:   env.Location(),
+		resourceID: resourceID,
+		converter:  converter,
+	}
+}
+
+type openShiftClusterStaticValidator struct {
+	location   string
+	resourceID string
+	converter  OpenShiftClusterConverter
+
+	r azure.Resource
+}
+
+// Validate validates an OpenShift cluster
+func (sv *openShiftClusterStaticValidator) Static(_oc interface{}, current *OpenShiftCluster) error {
```

Comment:
I didn't change the validation logic, but we now run it on the internal representation instead of external. As a result `Static` changed.

The specific API versions are still able to run validations on their (external) representation by wrapping this validator.

Ah, and `validateDelta` below changed a bit to handle `immutable.ValidationError`s and convert it into a `CloudError`.

Reply:
I was thinking we intentionally are running validation on the external model (a lesson from V3) ? @jim-minter ?

Reply:
Yes, on balance I think it's better to validate on external.  The error messages match the field names.  The down-side is duplication of the validation across different api versions.  I don't think bringing it internal can be the answer to that.  I don't know if there are other tactics we can employ to make this easier.

Reply:
I wonder what an approach using something like https://godoc.org/github.com/go-playground/validator might turn out like?  If done right it might get a lot of the easy stuff, e.g. "is this a valid url?" "is it an IP?", etc and just leave the fiddly ones (does the subscriptionid of this subnet match the subscriptionid of that subnet) and thus reduce the duplication of validating externally?

Reply:
Do we have any other concerns apart from non-matching names in errors?

I had a look at `github.com/go-playground/validator`. We can use the simple cases like "disk size is min=128" or "worker node count is min=3 and max=20". But most of the other validations will likely require custom validation functions. For example, to validate that resource ID in the URL matches one in the cluster object or VM sizes is one of the allowed. I think, we can build validation functions so they do not rely on specific type and can be reused across API versions, but it will be a not very intuitive solution it seems.

If the potential mismatch between internal and external field names is the only concern, maybe we should look into introducing an explicit converter between internal and external validation errors? We still can validate on external, if we need to do version-specific validation...

Reply:
I think we should just version our validators as we do with API's. We tried this in V3 (not to copy code around) and we failed. In the end we started copying code around for backwards compatibility. We will be doing the same for API's. let's do the same for Validators.

Reply:
@mjudeikis you suggest we close this PR and merge https://github.com/Azure/ARO-RP/pull/118 as is (in terms of validators)?

We should definitely be able to do version-sepcific validation (this PR enables this) becase later on we will be diverging from base validator. But I'm worried that with versioned validators it will be hard to keep validation in sync across versions: easy to make a change in one place and forget to do in another. Probably it will not be a problem once we have more stable API, but during active development - it can be.

But I don't know what is the less evil in this case - duplicated code in API versions or more complicated code with type conversions, etc...

Reply:
Reality is little bit different. When we create new API, we copy old one and make changes. Old API should never change, so old validator should not too. They should be frozen in time. We just decouple code logic. its code duplication, but enables us to move forward faster. It worked in v3, but we will need to automate code generation here to avoid pitfall we have with v3 releases.

Reply:
@mjudeikis Yes, that's why I said about active development phase. Once we stabilise the API I see only few cases when we might need to change API versions:

* When underlying packge such as (`pkg/env`, `pkg/util/subnet`, etc) or 3rd party package breaks compatibility
* When fixing a bug which affects multiple versions

So there is still risk of missing something, but how big is it? After having this discussion I less and less convinced that trying to get rid of code duplication in API versions worth it.

Reply:
me too. I personally think we can try to do so when we get into stable development cycle and we have stable apis and logic around them.